blog/fast-portfolio-sorts/index.qmd

```diff
@@ -0,0 +1,178 @@
+- Use the built-in `base` functions that ship with every R installation.
+- Leverage the popular `dplyr` package and workhorse of [Tidy Finance with R](../../r/index.qmd).
+- Explore the powerful `data.table` package and its sorting capabilities.
+- Combine the `dplyr` syntax with `data.table`'s performance through `dtplyr`.
+
+Throughout this blog post, I'll use the following packages. Notably, `bench` is used to create benchmarking results. 
+
+```{r}
+#| message: false
+#| warning: false
+library(dplyr)
+library(dtplyr)
+library(data.table)
+library(bench)
+library(purrr)
+library(RSQLite)
+```
+
+## Data preparation
+
+First, I start by loading the monthly CRSP data from our database (see [WRDS, CRSP, and Compustat](../../r/wrds-crsp-and-compustat.qmd) for details). The dataset has about 3 million rows and contains monthly returns between 1960 and 2023 for about 26,000 stocks. I also make sure that the data comes as a `tibble` for `dplyr`, a `data.frame` for `base`, a `data.table` for `data.table` and a “lazy” data table for `dtplyr` because I want to avoid any conversion issues in the portfolio assignments.
+
+```{r}
+tidy_finance <- dbConnect(
+  SQLite(),
+  "../../data/tidy_finance_r.sqlite",
+  extended_types = TRUE
+)
+
+crsp_monthly_dplyr <- tbl(tidy_finance, "crsp_monthly") |>
+  select(permno, month, ret_excess, mktcap_lag) |>
+  collect()
+
+crsp_monthly_base <- as.data.frame(crsp_monthly_dplyr)
+
+crsp_monthly_dt <- copy(as.data.table(crsp_monthly_dplyr))
+
+crsp_monthly_dtplyr <- lazy_dt(crsp_monthly_dplyr)
+```
+
+## Defining portfolio sorts
+
+As a common denominator across approaches, I introduce a stripped down version of `assign_portfolio()` that can also be found in the [`tidyfinance`](https://cran.r-project.org/web/packages/tidyfinance/index.html) package. 
+
+```{r}
+assign_portfolio <- function(data, sorting_variable, n_portfolios) {
+  
+  breakpoints <- quantile(
+    data[[sorting_variable]], 
+    probs = seq(0, 1, length.out = n_portfolios + 1), 
+    na.rm = TRUE, names = FALSE
+  )
+
+  findInterval(
+    data[[sorting_variable]], breakpoints, all.inside = TRUE
+  )
+}
+```
+
+The goal is to apply this function to the cross-section of stocks in each month and then compute average excess returns for each portfolio across all months.  
+
+If we want to apply the function above to each month using only `base`, then we have to first split the `data.frame` into multiple parts and `lapply()` the function to each part. After we combined the parts again to one big data.frame, we can use `aggregate()` to compute the average excess returns. 
+
+```{r}
+sort_base <- function() {
+  df_split <- split(crsp_monthly_base, crsp_monthly_base$month)
+  df_split <- lapply(df_split, function(df) {
+    df$portfolio <- assign_portfolio(df, "mktcap_lag", n_portfolios = 10)
+    return(df)
+  })
+  df <- do.call(rbind, df_split)
+  aggregate(ret_excess ~ portfolio, data = df, FUN = function(x) mean(x, na.rm = TRUE))
+}
+sort_base()
+```
```

Comment:
I am not familiar with aggregate but if this replaces a group_by(portfolio) |> summarise() call, can't it also be used instead of the lapply part? 
Also: is it the split, the lapply, or the do.call that is so slow? Unlist typically also works on lapply lists and should do something alike than do.call(rbind,..). If it is the aggregate which causes trouble, then there are of course other ways to compute the mean, e.g. via lapply (to sum per split and then sum all list elements).

Reply:
Nah, aggregate only computes summary statistics and is fairly fast, but the lapply part only adds a new column. The slowest part is the do.call(rbind, df_split), followed by  split(crsp_monthly_base, crsp_monthly_base$month). I can't come up with a better solution at the moment..

Reply:
What's more troubling is that the resulting portfolio returns are different from the other approaches and I can't figure out why

Reply:
This is actually considerably faster, but still different results.. 
```
crsp_monthly_base$portfolio <- with(
    crsp_monthly_base, 
    ave(mktcap_lag, month, FUN = function(x) assign_portfolio(data.frame(mktcap_lag = x), "mktcap_lag", n_portfolios = 10))
  )
  
  mean_ret_excess <- with(
    crsp_monthly_base, 
    tapply(ret_excess, portfolio, function(x) mean(x, na.rm = TRUE))
  )

  data.frame(
    portfolio = names(mean_ret_excess), 
    ret = unlist(mean_ret_excess)
  )
```